### PR TITLE
Release GIL while awaiting topology compile

### LINKED ISSUE #38829: `register_hlo_module_transformation` can deadlock on TPU `CompileWithTopology` because the Python transform callback needs the GIL
### Description

When using `jax.extend.xla.register_hlo_module_transformation(...)` for a TPU/plugin backend, a Python HLO transform callback can deadlock during topology-based AOT compilation.

The symptom is that compilation hangs after XLA reaches the registered pre-scheduler transform. The Python callback body is never entered: even the first log line in the callback does not execute.

The suspected cause is a GIL ownership cycle:

1. The Python main thread starts a topology-based compile path that reaches `PyClient::CompileWithTopology`.
2. `CompileWithTopology` blocks in `Compile(...).Await()`.
3. For plugin backends, `register_hlo_module_transformation` registers the Python callback through the PJRT C API transform extension.
4. XLA invokes that C API transform callback synchronously from the compiler pipeline.
5. jaxlib's C API bridge tries to acquire the Python GIL before calling the Python callback.
6. If the Python main thread still holds the GIL while blocked in `CompileWithTopology().Await()`, the compiler thread cannot enter the callback, while the main thread is waiting for compilation to finish.

This makes Python HLO transforms unusable in this TPU AOT path.

## Relevant code paths

JAX public registration path:

- `jax/_src/xla_transform.py`
- `register_hlo_module_transformation(...)`
- For plugin backends it calls `_xla.register_xla_transform_c_api(client, name, stage_int, callback)`.
- Upstream main around lines 80-85:

```python
def register_on_client(client):
  if client.platform in platforms_list:
    try:
      _xla.register_xla_transform_c_api(client, name, stage_int, callback)
```

JAX C API bridge reacquires the GIL:

- `jaxlib/xla.cc`
- `CApiTransformHloModuleCallback(...)`
- Upstream main around lines 130-146:

```c++
void CApiTransformHloModuleCallback(PJRT_XlaTransform_Callbacks* callbacks,
                                    PJRT_XlaTransform_Args* args) {
  ...
  nb::gil_scoped_acquire gil;
  try {
    nb::bytes input_bytes(args->hlo_module.data, args->hlo_module.size);
    nb::object result = state->py_callback(input_bytes);
```

Topology-based compile path appears to wait without releasing the GIL:

- `jaxlib/py_client.cc`
- `CompileWithTopology(...)`
- Upstream main around lines 434-450:

```c++
static absl::StatusOr<nb_class_ptr<PyExecutable>> CompileWithTopology(
    nb_class_ptr<PyClient> client, mlir::ModuleOp module,
    const xla::ifrt::Topology& topology, xla::CompileOptions options,
    ifrt::DeviceListRef devices) {
  mlir::OwningOpRef<mlir::ModuleOp> clone(module.clone());
  options.allow_in_place_mlir_modification = true;
  ifrt::ExecutableRef ifrt_executable;
  {
    auto xla_options =
        std::make_unique<ifrt::XlaCompileOptions>(options, std::move(devices));
    TF_ASSIGN_OR_RETURN(
        ifrt_executable,
        client->ifrt_client()
            ->GetDefaultCompiler()
            ->Compile(std::make_unique<xla::ifrt::HloProgram>(std::move(clone)),
                      topology, std::move(xla_options))
            .Await());
  }
```

XLA invokes the registered transform synchronously:

- `xla/pjrt/c/pjrt_c_api_xla_transform_internal.cc`
- `CApiXlaTransformAdapter::Transform(...)`
- Around lines 62-80 in the local XLA checkout:

```c++
absl::StatusOr<bool> Transform(xla::HloModule* module) override {
  xla::HloModuleProto proto = module->ToProto();
  ...
  if (callbacks_->transform_hlo_module == nullptr) {
    return absl::InternalError("transform_hlo_module callback is null");
  }
  callbacks_->transform_hlo_module(callbacks_, &args);
```

The transform is called by the HLO pass:

- `xla/service/xla_transform.cc`
- `ApplyXlaTransformsToModule(...)`
- Around lines 101-116 in the local XLA checkout:

```c++
absl::StatusOr<bool> ApplyXlaTransformsToModule(
    HloXlaTransform::PipelineStage stage, xla::HloModule* module) {
  ...
  for (auto& transform : transforms) {
    auto status_or_bool = transform->Transform(module);
```

## Observed behavior

With a registered no-op Python HLO transform:

```python
from jax.extend import xla

def callback(hlo_module_proto: bytes):
  print("callback_start", flush=True)
  return None

xla.register_hlo_module_transformation(
    "noop_hlo_transform",
    callback,
    stage=xla.PipelineStage.PRE_SCHEDULER,
    platforms="tpu",
)
```

Topology-based TPU AOT compilation hangs. The `callback_start` line is never printed.

The native stack observed during the hang shows the compiler thread in the transform path:

```text
CApiTransformHloModuleCallback
CApiXlaTransformAdapter::Transform
ApplyXlaTransformsToModule
ApplyXlaTransforms::RunImpl
RunHloScheduler / pre-scheduler pipeline
```

The Python/main thread is blocked in the topology compile path waiting on:

```text
PyClient::CompileWithTopology
GetDefaultCompiler()->Compile(...).Await()
```

## Expected behavior

Registering a Python HLO transform for a TPU/plugin backend should not deadlock during topology-based compilation. The callback should either run successfully or propagate a Python/C++ error back to the compile call.

## Possible fix

`CompileWithTopology` should release the Python GIL while waiting for the IFRT compile future, similar to other blocking jaxlib compile/execute paths.

For example:

```c++
{
  nb::gil_scoped_release gil_release;
  auto xla_options =
      std::make_unique<ifrt::XlaCompileOptions>(options, std::move(devices));
  TF_ASSIGN_OR_RETURN(
      ifrt_executable,
      client->ifrt_client()
          ->GetDefaultCompiler()
          ->Compile(std::make_unique<xla::ifrt::HloProgram>(std::move(clone)),
                    topology, std::move(xla_options))
          .Await());
}
```

This matches the behavior of `CompileOnlyPyClient::CompileUnloaded`, which already releases the GIL around the compile call.

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.10.2
jaxlib: 0.10.2
numpy:  2.4.6
python: 3.12.12 (main, Feb  4 2026, 20:22:22) [GCC 12.2.0]
device info: TPU7x-8, 8 local devices"
process_count: 1
platform: uname_result(system='Linux', node='gke-tpu-008e4da3-xphw', release='6.12.68+', version='#1 SMP Sun Apr 12 08:53:59 UTC 2026', machine='x86_64')

Additional context from the observed reproduction:

- Backend: TPU plugin backend
- Compilation mode: topology-based AOT compile path
- Transform stage: `PipelineStage.PRE_SCHEDULER`
- Transform registration: `jax.extend.xla.register_hlo_module_transformation(..., platforms="tpu")`

### PR DESCRIPTION
**Pull request by hhhhsdxxxx**
Fixes #38829.

`CompileWithTopology` waits on an IFRT compile future via `Compile(...).Await()`. When the future invokes Python HLO module transformation callbacks, those callbacks need to acquire the Python GIL. Holding the GIL while awaiting the future can deadlock the callback path.

This releases the GIL around the blocking topology compile await, matching the pattern used by other blocking compile paths in `py_client.cc`.

Tested manually in a TPU AOT compile path with a Python HLO transformation callback: the callback enters Python and the compile completes successfully.

### CODE REVIEW

**Comment by hhhhsdxxxx** [comment]
Hi @hawkinsp, friendly ping — when you have a chance, could you please take a look at this PR and approve the pending workflows (or route it to the appropriate reviewer)? Thanks!

**Comment by hhhhsdxxxx** [comment]
Closing this PR because the reported deadlock has been indirectly eliminated in current upstream by OpenXLA commit [329d892](https://github.com/openxla/xla/commit/329d8923474b880c43940b40ff8d99d29172fcea), "Disable async compilation in PjRt-IFRT".

JAX 0.10.2 pinned XLA at `5a9e73c`. In that version, TPU `PjRtCompiler::Compile` submitted `PjRtExecutable::Create` to an eight-thread `PjRtCompilerThreadPool` through `MakeFutureOn`. The Python caller retained the GIL in `CompileWithTopology().Await()`, while the compiler worker needed the GIL to enter the Python HLO transform callback, creating the deadlock cycle.

The OpenXLA change removed that compilation thread pool and now calls `PjRtExecutable::Create` directly on the caller thread. As a result, the HLO transform callback is reached on the GIL-owning thread and its nested GIL acquisition succeeds; the returned future is already resolved when JAX calls `.Await()`.

This is not dependent on the later libtpu 0.0.44 update. A TPU7x CI run using libtpu 0.0.42.1 and a regular GIL-enabled Python 3.12 passed `//tests:xla_transform_test_tpu`: https://github.com/jax-ml/jax/actions/runs/29124621633/job/86469201893

The explicit `gil_scoped_release` in this PR would still be a more structural safeguard and would avoid holding the GIL throughout a long topology compile. However, the original deadlock no longer reproduces with the current synchronous PjRt-IFRT compilation path, so I am closing this PR. This should be revisited if asynchronous compilation is reintroduced or a PJRT plugin invokes the transform callback from a worker thread.
